src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/CheckstyleJavaParser.g4

```diff
@@ -0,0 +1,858 @@
+    | LPAREN formalParameterList? RPAREN                                   #formalLambdaParam
+    | LPAREN multiLambdaParams RPAREN                                      #multiLambdaParam
+    ;
+
+multiLambdaParams
+    : id (COMMA id)*
+    ;
+
+// Java8
+lambdaBody
+    : expression
+    | block
+    ;
+
+primary
+    : switchExpressionOrStatement                                          #switchPrimary
+    | LPAREN expr RPAREN                                                   #parenPrimary
+    | LITERAL_THIS                                                         #tokenPrimary
+    | LITERAL_SUPER                                                        #tokenPrimary
+    | literal                                                              #literalPrimary
+    | id                                                                   #tokenPrimary
+    | type=primaryClassOrInterfaceTypeNoArray arrayDeclarator*
+      DOT LITERAL_CLASS                                                    #classRefPrimary
+    | type=primaryPrimitiveTypeNoArray arrayDeclarator*
+      DOT LITERAL_CLASS                                                    #primitivePrimary
+    | nonWildcardTypeArguments LITERAL_THIS arguments                      #paramPrimary
+    ;
+
+primaryClassOrInterfaceTypeNoArray
+    : annotations[false] classOrInterfaceType[false]
+    ;
+
+primaryPrimitiveTypeNoArray
+    : annotations[false] primitiveType
+    ;
+
+classType
+    : (classOrInterfaceType[false] DOT)? annotations[false] id typeArguments?
+    ;
+
+creator
+    : nonWildcardTypeArguments createdName classCreatorRest
+    | annotations[false] createdName
+      (annotations[false] arrayCreatorRest | classCreatorRest)
+    ;
+
+createdName
+    : id typeArgumentsOrDiamond? extended+=createdNameExtended*            #createdNameObject
+    | primitiveType                                                        #createdNamePrimitive
+    ;
+
+createdNameExtended
+    : DOT annotations[false] id typeArgumentsOrDiamond?
+    ;
+
+innerCreator
+    : annotations[false] id nonWildcardTypeArgumentsOrDiamond? classCreatorRest
+    ;
+
+arrayCreatorRest
+    : LBRACK
+      ( RBRACK arrayDeclarator* arrayInitializer
+      | expression RBRACK bracketsWithExp* arrayDeclarator*
+      )
+    ;
+
+bracketsWithExp
+    : annotations[false] LBRACK expression RBRACK
+    ;
+
+classCreatorRest
+    : arguments classBody?
+    ;
+
+typeArgumentsOrDiamond
+    : LT GT                                                                #diamond
+    | typeArguments                                                        #typeArgs
+    ;
+
+nonWildcardTypeArgumentsOrDiamond
+    : LT GT                                                                #nonWildcardDiamond
+    | nonWildcardTypeArguments                                             #nonWildcardTypeArgs
+    ;
+
+nonWildcardTypeArguments
+    : LT typeArgumentsTypeList GT
+    ;
+
```

Comment:
Please add comment that explains why we need both `typeArgumentsTypeList` and `typeList`

Reply:
I did this to be able to handle the same construct differently in the visitor, while keeping AST consistent with current Checkstyle structure:

```java

    @Override
    public DetailAstImpl visitTypeArgumentsTypeList(
            CheckstyleJavaParser.TypeArgumentsTypeListContext ctx) {
        final DetailAstImpl firstIdent = visit(ctx.typeType(0));
        final DetailAstImpl firstTypeArgument = createImaginary(TokenTypes.TYPE_ARGUMENT);
        firstTypeArgument.addChild(firstIdent);

        for (int i = 0; i < ctx.COMMA().size(); i++) {
            addLastSibling(firstTypeArgument, create(ctx.COMMA(i)));
            final DetailAstImpl ident = visit(ctx.typeType(i + 1));
            final DetailAstImpl typeArgument = createImaginary(TokenTypes.TYPE_ARGUMENT);
            typeArgument.addChild(ident);
            addLastSibling(firstTypeArgument, typeArgument);
        }
        return firstTypeArgument;
    }

    @Override
    public DetailAstImpl visitTypeList(CheckstyleJavaParser.TypeListContext ctx) {
        return flattenedTree(ctx);
    }

```

I could also re-use `typeList` and handle both cases in one `visit...` method. When this decision came up, I usually chose to keep `visit...` methods as simple as possible, as long as it did not create ambiguous grammar. Do you think this was a good design decision?

Reply:
I would only like to comment that these expressions build different ASTs, so they are duplicated so that whoever will make changes to this file in a few years will have more information.

Reply:
Comment added.